File: backend/app/schemas/analytics.py
"""
Analytics schemas — for authority dashboard and reporting.
"""
from typing import Optional, Any, Dict, List
from uuid import UUID
from datetime import datetime, date
from pydantic import BaseModel, ConfigDict


class DashboardSummary(BaseModel):
    """Top-level metrics for authority dashboard."""
    total_incidents: int
    total_near_misses: int
    active_incidents: int
    verified_incidents: int
    resolved_incidents: int
    rejected_incidents: int
    high_severity_count: int
    medium_severity_count: int
    low_severity_count: int
    high_risk_zone_count: int
    avg_risk_score: Optional[float]


class HotspotData(BaseModel):
    """Single data point for heatmap rendering."""
    latitude: float
    longitude: float
    weight: float  # 0.0 - 1.0 normalized intensity


class HotspotResponse(BaseModel):
    heatmap_points: List[HotspotData]
    top_risk_zones: List[Dict[str, Any]]
    incident_type_breakdown: Dict[str, int]
    hazard_type_breakdown: Dict[str, int]


class TrendPoint(BaseModel):
    date: date
    incident_count: int
    near_miss_count: int


class AnalyticsSnapshotResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    snapshot_date: date
    total_incidents: int
    total_near_misses: int
    active_incidents: int
    verified_incidents: int
    resolved_incidents: int
    high_severity_count: int
    avg_risk_score: Optional[float]
    top_risk_zones: Optional[List[Dict[str, Any]]]
    incident_type_breakdown: Optional[Dict[str, int]]
    created_at: datetime

File: frontend/src/types/analytics.ts
// Risk zone and analytics TypeScript types

export interface RiskZone {
  id: string
  name: string
  description: string | null
  center_latitude: number
  center_longitude: number
  radius_meters: number
  risk_score: number
  accident_count: number
  near_miss_count: number
  high_severity_count: number
  contributing_factors: Record<string, unknown> | null
  last_calculated_at: string | null
  created_at: string
}

export interface RiskZoneListItem {
  id: string
  name: string
  center_latitude: number
  center_longitude: number
  radius_meters: number
  risk_score: number
  accident_count: number
  near_miss_count: number
}

export interface RouteWarning {
  risk_zone_id: string
  name: string
  risk_score: number
  reasons: string[]
  center_latitude: number
  center_longitude: number
  suggested_action: string
}

export interface DashboardSummary {
  total_incidents: number
  total_near_misses: number
  active_incidents: number
  verified_incidents: number
  resolved_incidents: number
  rejected_incidents: number
  high_severity_count: number
  medium_severity_count: number
  low_severity_count: number
  high_risk_zone_count: number
  avg_risk_score: number | null
}

export interface HotspotPoint {
  latitude: number
  longitude: number
  weight: number
}

export interface HotspotResponse {
  heatmap_points: HotspotPoint[]
  top_risk_zones: RiskZoneListItem[]
  incident_type_breakdown: Record<string, number>
  hazard_type_breakdown: Record<string, number>
}

export interface TrendPoint {
  date: string
  incident_count: number
  near_miss_count: number
}

// Emergency services types
export interface EmergencyService {
  name: string
  type: 'hospital' | 'police' | 'trauma_center'
  latitude: number
  longitude: number
  distance_km: number
  phone: string | null
  address: string | null
}

// API pagination wrapper
export interface PaginatedResponse<T> {
  success: boolean
  data: T[]
  message: string
  pagination: {
    page: number
    limit: number
    total: number
  }
}
